support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala

```diff
@@ -1,78 +1,242 @@
+      gifter_first_name = user.firstName,
+      gifter_last_name = user.lastName,
+      gift_recipient_first_name = "gift recipient first name placeholder",
+      gift_recipient_last_name = "gift recipient last name placeholder",
+      gift_recipient_email = "gift recipient email placeholder",
+      gift_personal_message = "gift personal message placeholder",
+      gift_code = "gift code placeholder",
+      gift_delivery_date = "gift delivery date placeholder",
+      subscription_details = "subscription details placeholder",
+      date_of_first_payment = "date_of_first_payment",
+      paymentAttributes = paymentFields(pm, directDebitMandateId)
+    )
+
+  private def buildGiftRedemptionAttributes =
+    GifteeRedemptionAttributes(
+      gift_recipient_first_name = user.firstName,
+      subscription_details = "subscription_details placeholder",
+      gift_start_date = "gift start date placeholder",
+      gift_recipient_email = user.primaryEmailAddress
+    )
+
+  private def buildCorpRedemptionAttributes =
+    directOrCorpFields("Group subscription")
+
+  private def directAttributes(pm: PaymentMethod, paymentSchedule: PaymentSchedule) =
+    DirectDSAttributes(
+      directOrCorpFields(SubscriptionEmailFieldHelpers.describe(paymentSchedule, billingPeriod, currency, promotion)),
+      subscription_term = billingPeriod.noun,
+      payment_amount = SubscriptionEmailFieldHelpers.formatPrice(SubscriptionEmailFieldHelpers.firstPayment(paymentSchedule).amount),
+      country = user.billingAddress.country.name,
+      date_of_first_payment = formatDate(SubscriptionEmailFieldHelpers.firstPayment(paymentSchedule).date),
+      currency = currency.glyph,
+      trial_period = "14", //TODO: depends on Promo code or zuora config
+      paymentFields(pm, directDebitMandateId)
+    )
+
+}
+
+object DigitalPackEmailFields {
+
+  type Failable[A] = Either[String, A]
+  def asFlattenedPairs(value: JsonObject): Failable[List[(String, String)]] = {
+    def flattenToPairs(value: JsonObject, failablePairsSoFar: Failable[Map[String, String]]): Failable[Map[String, String]] =
+      value.toList.foldLeft(failablePairsSoFar) {
+          case (failablePairsSoFar, (fieldName, jValue)) =>
+            failablePairsSoFar.flatMap { pairsSoFar =>
+              jValue.asString match {
+                case None =>
+                  jValue.asObject match {
+                    case None => Left(s"all values should be string or object: ${fieldName} -> $value")
+                    case Some(obj) =>
+                      flattenToPairs(obj, Right(pairsSoFar))
+                  }
+                case Some(string) =>
+                  if (pairsSoFar.contains(fieldName))
+                    Left(s"found duplicate key ${fieldName} in case classes")
+                  else
+                    Right(pairsSoFar + (fieldName -> string))
+              }
+          }
+        }
+
+    flattenToPairs(value, Right(Map.empty)).map(_.toList)
   }
```

Comment:
this just turns a JsonObject into a List of key/value pairs for insertion into the existing EmailFields object.  We don't want to change the non DS emails at this point, so it makes sense to convert the case classes to the existing structure.

Reply:
This is useful information include it in a comment?

Reply:
good idea, will do.  I might even pull that function out to a separate file, as it's nothing to do with emails really, it's just a simple conversion function.